### PR TITLE
[16.0][IMP] fieldservice_geoengine: Fix vector Layer related Errors

### PR DESCRIPTION
**Pull request by anusriNPS**
- [ ] Depends on base_geoengine PR [#417](https://github.com/OCA/geospatial/pull/417#top)

Domain for "attribute_field_id" is filtered based on "classification" field selection in vector layer, which helps to avoid JS error seen as "fsm.order" and "fsm.location" related fields are not validated based on "NUM_ATT" types.[ base_geoengine_validation](https://github.com/OCA/geospatial/blob/16.0/base_geoengine/models/geo_vector_layer.py#L122).

   Notifying user to define selected field from attribute_field_id of supported type in geoengine view which avoids below JS error observed:
   Screenshot from 2025-10-08 17-09-18

   Notifying users when a field of supported type from attribute_field_id is not suitable to use for classification type "custom" which avoids below observed JS Error:
   image
   
  As  fieldservice_geoengine omits validation for custom classification related to "type" check on field selected in "attribute_field_id", when user selects fields of type "date" with "custom", below JS error observed:
   image

### CODE REVIEW

**Comment by anusriNPS** [comment]
Notifying user to define selected field in geoengine view from attribute_field_id whose field is of supported type:
Screenshot from 2025-10-13 08-49-45


Notifying user when selected field in "attribute_field_id" is of supported type but not suitable for interval/custom range. Example: "color" field which is of "integer" type in "fsm.location" but could not generate interval/custom range, if not suppressed JS error is seen.

Interval range:
image
Custom range:
image


For custom classification with field type such as "date", based on this PR it suppress JS error but view displayed for "custom" will not be the expected view. This is because user needs to use proper field for selected field type to display customized view. Below is the view when "request_early" is selected from "attribute_field_id" for "custom".

image

File: fieldservice_geoengine/models/vector_layer.py [hunk 1]
```diff
@@ -7,6 +7,18 @@
...
+    "integer",
+    "integer_big",
+    "related",
+    "function",
+    "date",
+    "datetime",
+    "char",
+    "text",
+    "selection",
+]
```
Comment (HekkiMelody) [inline fieldservice_geoengine/models/vector_layer.py:21]: **chore** (non-blocking): I see these elsewhere in the code as well, but I suspect they may have remained from previous versions, as I cannot find values of `ttype` matching some of these.

On a fresh odoo installation, opening a shell and inspecting `env["ir.model.fields"]._fields["ttype"].selection` I get the following fields

```
['binary', 'boolean', 'char', 'date', 'datetime', 'float', 'html', 'integer', 'json', 'many2many', 'many2one', 'many2one_reference', 'monetary', 'one2many', 'properties', 'properties_definition', 'reference', 'selection', 'text', 'serialized']
```

File: fieldservice_geoengine/models/vector_layer.py [hunk 2]
```diff
@@ -31,3 +43,27 @@ def _check_geo_repr(self):
...
+    @api.depends("geo_field_id", "classification", "geo_repr")
+    def _compute_attribute_field_id_domain(self):
+        for rec in self:
+            if rec.geo_field_id:
+                if (
+                    rec.geo_repr == "colored"
+                    and (
+                        rec.classification != "unique"
+                        and rec.classification != "custom"
+                    )
```
Comment (HekkiMelody) [inline fieldservice_geoengine/models/vector_layer.py]: **suggestion** (non-blocking): readability


```suggestion
                    rec.geo_repr == "colored"
                    and rec.classification not in ("unique", "custom")
```

I would also add  brackets for clarity

```py
                    (rec.geo_repr == "colored"
                    and rec.classification not in ("unique", "custom"))
                    or ....
```

**Review by HekkiMelody — APPROVED** [review]
Code review, LGTM

File: fieldservice_geoengine/models/vector_layer.py (hunk 1 above)
Comment (anusriNPS) [reply fieldservice_geoengine/models/vector_layer.py:21]: `env["ir.model.fields"]._fields["ttype"].selection` - list all available `ttype` values.

`fieldservice_geoengine` supports only fields based on `SUPPORTED_ATT` and `NUM_ATT` which is used in domain to filter from `ttype` values.

`ttype` will not match the fields of `supported_att` and `num_att` variables and it should list all available types. `ttype` list seems right and this is my understanding. Please correct me, if my views differ from your thoughts.

File: fieldservice_geoengine/models/vector_layer.py (hunk 2 above)
Comment (anusriNPS) [reply fieldservice_geoengine/models/vector_layer.py]: Updated as suggested

File: fieldservice_geoengine/models/vector_layer.py (hunk 1 above)
Comment (HekkiMelody) [reply fieldservice_geoengine/models/vector_layer.py:21]: Sorry, I probably wasn't very clear. I didn't mean to imply that fields should be added to `SUPPORTED_ATT`, but rather that some may be removed (maybe?), like `integer_big`, `related` and `function`.

Either way, non-blocking.

**Review by HekkiMelody — APPROVED** [review]

**Comment by HekkiMelody** [comment]
~~@OCA/geospatial-maintainers~~ Hello, could you take a look? Thanks
Wrong ping, sorry, I meant @OCA/field-service-maintainers

**Comment by hparfr** [comment]
Does it need https://github.com/OCA/geospatial/pull/417 to be merged first ?

**Comment by HekkiMelody** [comment]
> Does it need [OCA/geospatial#417](https://github.com/OCA/geospatial/pull/417) to be merged first ?

yes, actually, it does

**Review by marcos-mendez — CHANGES_REQUESTED** [review]
**Automated Review -- Tests Failed**

### 1. **Root Cause**

The test failure is caused by a **database connection error** during the Odoo startup, not directly related to the changes in the PR. However, the new `oca_dependencies.txt` file references a **nonexistent or broken geospatial dependency**, which likely leads to a failure in module loading or dependency resolution.

### 2. **Suggested Fix**

- **File:** `fieldservice_geoengine/oca_dependencies.txt`
- **Issue:** The dependency URL and commit hash are invalid or unreachable.
- **Fix:** Either:
  - Correct the dependency URL or commit hash to a valid reference.
  - Remove the dependency if it's not required for the current changes.
  - Confirm that the referenced PR (`refs/pull/417/head`) is stable and accessible.

> **Line:** 1  
> **Current:** `geospatial https://github.com/OCA/geospatial refs/pull/417/head 9825f67d128918294b67b74d324d1e6d3259399b`  
> **Suggestion:** Update to a working commit or remove if unnecessary.

---

### 3. **Additional Code Issues**

#### a. **Potential crash in JS code**
- **File:** `fieldservice_geoengine/static/src/js/views/geoengine/geoengine_renderer/geoengine_renderer.esm.js`
- **Issue:** In `styleVectorLayer` method, the `try...catch` block silently swallows errors and returns early without logging or notifying the user.
- **Fix:** Log the error or provide a more informative fallback.

> **Lines:** 113–118  
> **Suggestion:** Add `console.error(error)` or a more user-facing error notification.

#### b. **Inconsistent handling of `attribute_field_id_domain`**
- **File:** `fieldservice_geoengine/models/vector_layer.py`
- **Issue:** The `_compute_attribute_field_id_domain` method builds a domain based on conditions, but it does not validate that `rec.geo_field_id.model_id.model` is a valid model.
- **Fix:** Add a check or error handling to ensure `geo_field_id.model_id.model` is valid before using it in a domain.

> **Lines:** 48–56  
> **Suggestion:** Add a `try-except` or validation for `rec.geo_field_id.model_id.model`.

---

### 4. **Test Improvements**

#### a. **Add test for `oca_dependencies.txt`**
- **Type:** `SavepointCase` or `TransactionCase`
- **Test Case:** Ensure the module loads without errors when `oca_dependencies.txt` is present and valid.
- **Tags:** `@odoo-test`, `@runboat-ready`

#### b. **Test `attribute_field_id_domain` computation**
- **Type:** `TransactionCase`
- **Test Case:** Create a `GeoVectorLayer` record and assert that `attribute_field_id_domain` is correctly computed based on `geo_repr` and `classification`.
- **Tags:** `@odoo-test`, `@geoengine`

#### c. **Test JS error handling in `geoengine_renderer.esm.js`**
- **Type:** `SavepointCase` or `HttpCase`
- **Test Case:** Simulate an invalid `label_text` to ensure the JS code doesn't crash and logs appropriately.
- **Tags:** `@odoo-test`, `@frontend`, `@geoengine`

#### d. **Test `checkAttributeFieldUsage` and `getClass` methods**
- **Type:** `TransactionCase` or `SavepointCase`
- **Test Case:** Test edge cases in `getClass` like `null`, `undefined`, or invalid values.
- **Tags:** `@odoo-test`, `@geoengine`

---

### Summary

- The **test failure is likely due to a broken dependency**, not the core logic.
- The **JS code has a silent error handler** that should be improved.
- The **domain computation logic** should include validation for model references.
- Add **unit tests** for domain logic, JS error handling, and frontend behavior.

---
### ⏰ PR Aging Alert

This PR by @anusriNPS has been open for **158 days** (5 months).

> _Every ignored PR is a contributor who might not come back. Review time matters. ([OCA Aging Report](https://git.pop.coop/pop/oca-reviewer))_

---
### Reciprocal Review Request



Hi everyone! I found some test failures on this PR and left detailed feedback above. I am happy to discuss or help debug. In the meantime, if any of you get a chance, I would appreciate a look at my open PR(s):

**My open PRs across OCA:**
- [server-tools#3554 [MIG] datetime_formatter: Migration to 18.0](https://github.com/OCA/server-tools/pull/3554)
- [server-tools#3548 [18.0][MIG] base_kanban_stage: Migration to 18.0](https://github.com/OCA/server-tools/pull/3548)
- [hr-attendance#262 [16.0][ADD] Hr_attendance_idsecure: iDSecure (ControliD) attendance integration](https://github.com/OCA/hr-attendance/pull/262)
- [stock-logistics-workflow#2276 [16.0][ADD] stock_move_line_devaluation](https://github.com/OCA/stock-logistics-workflow/pull/2276)
- [stock-logistics-workflow#2275 [16.0][ADD] Stock move line analytic account](https://github.com/OCA/stock-logistics-workflow/pull/2275)
- [stock-logistics-workflow#2268 [16.0][ADD] stock_move_line_picking_partner](https://github.com/OCA/stock-logistics-workflow/pull/2268)
- [purchase-workflow#2694 [16.0][IMP]Purchase workflow added to review state & exception fix](https://github.com/OCA/purchase-workflow/pull/2694)

Reviewing each other's work helps the whole community move forward. Thank you!

---
_Environment via [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer): Minikube + K8s Job + oca-ci/py3.10-odoo16.0 | Odoo 16.0_
_Automated review by [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer) + qwen3-coder:30b_

**Review by quirino95 — APPROVED** [review]
Code and functional review: LGTM!

**Review by marcos-mendez — APPROVED** [review]
Apologies for the noise from the automated review — the bot incorrectly used "changes requested" status, which revoked the approved label had already granted. I'm fixing the bot to use "comment" status instead, so this won't happen again.

**Review by hparfr — APPROVED** [review]
/ocabot merge minor
